### PR TITLE
Integrate ankr_getAccountBalance for querying balances

### PR DESCRIPTION
**Pull request by onyb**
Resolves https://github.com/brave/brave-browser/issues/33330

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Balance queries should be working as expected. Checklist of things to check:
- Portfolio view (verify all "Group by" options - None, Accounts, Networks)
- Account view
- Asset view
- Send token lists
- Swap token lists

### CODE REVIEW

File: components/brave_wallet/common/brave_wallet.mojom [hunk 1]
```diff
@@ -1349,6 +1360,9 @@ interface JsonRpcService {
                       string chain_id) => (array<SPLTokenAmount> balances,
                                            SolanaProviderError error,
                                            string error_message);
+
+  AnkrGetAccountBalance(string account, array<string> chain_ids)
```
Comment (StephenHeaps) [inline components/brave_wallet/common/brave_wallet.mojom:1364]: nit: Should this be `AnkrGetAccountBalances` (plural) as it's fetching multiple balances?
And maybe the `string account` should be `string account_address` for clarity?

File: components/brave_wallet_ui/common/slices/api.slice.ts [hunk 2]
```diff
@@ -136,23 +136,15 @@ type GetTokenBalancesForAddressAndChainIdArg = {
...
 
 type GetTokenBalancesRegistryArg = {
-  accountId: BraveWallet.AccountId,
-  chainId: string,
-  coin:
-        | typeof CoinTypes.ETH
-    | typeof CoinTypes.SOL
-    | typeof CoinTypes.FIL
-    | typeof CoinTypes.BTC
+  accounts: BraveWallet.AccountId[]
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/slices/api.slice.ts:146]: nit: can we rename this to `accountIds` for clarity?

File: components/brave_wallet/browser/json_rpc_requests_helper.cc [hunk 3]
```diff
@@ -76,4 +76,29 @@ base::flat_map<std::string, std::string> MakeCommonJsonRpcHeaders(
...
 
+base::flat_map<std::string, std::string> MakeBraveServicesKeyHeaders() {
+  return {
+      {kBraveServicesKeyHeader, BUILDFLAG(BRAVE_SERVICES_KEY)},
+  };
+}
+
+std::string EncodeAnkrGetAccountBalancesParams(
+    const std::string& address,
+    std::vector<std::string> blockchains) {
```
Comment (supermassive) [inline components/brave_wallet/browser/json_rpc_requests_helper.cc:87]: const &

File: components/brave_wallet/common/brave_wallet.mojom [hunk 4]
```diff
@@ -1359,6 +1371,9 @@ interface JsonRpcService {
                       string chain_id) => (array<SPLTokenAmount> balances,
                                            SolanaProviderError error,
                                            string error_message);
+
+  AnkrGetAccountBalances(string account_address, array<string> chain_ids)
+    => (array<AnkrAssetBalance> balances, string error_message);
 };
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet.mojom:1377]: Brief comment please

File: components/brave_wallet/common/features.h [hunk 5]
```diff
@@ -26,6 +26,7 @@ BASE_DECLARE_FEATURE(kBraveWalletSnsFeature);
 BASE_DECLARE_FEATURE(kBraveWalletBitcoinFeature);
 extern const base::FeatureParam<int> kBitcoinRpcThrottle;
 BASE_DECLARE_FEATURE(kBraveWalletZCashFeature);
+BASE_DECLARE_FEATURE(kBraveWalletAnkrBalancesFeature);
```
Comment (supermassive) [inline components/brave_wallet/common/features.h:29]: Maybe also worth adding to brave://flags ? 
brave/browser/about_flags.cc

File: components/brave_wallet/browser/json_rpc_service.cc [hunk 6]
```diff
@@ -3414,4 +3429,61 @@ void JsonRpcService::OnGetSPLTokenBalances(
                           mojom::SolanaProviderError::kSuccess, "");
 }
 
+void JsonRpcService::AnkrGetAccountBalances(
+    const std::string& account,
+    const std::vector<std::string>& chain_ids,
+    AnkrGetAccountBalancesCallback callback) {
+  auto internal_callback =
```
Comment (supermassive) [inline components/brave_wallet/browser/json_rpc_service.cc:3436]: Let's immediately fail if feature is disabled

File: components/brave_wallet/browser/brave_wallet_constants.cc [hunk 7]
```diff
@@ -108,4 +108,25 @@ const std::string GetAssetRatioBaseURL() {
...
+                   {mojom::kPolygonMainnetChainId, "polygon"},
+                   {mojom::kPolygonZKEVMChainId, "polygon_zkevm"},
+                   {mojom::kRolluxMainnetChainId, "rollux"},
+                   {mojom::kSyscoinMainnetChainId, "syscoin"},
+                   {mojom::kZkSyncEraChainId, "zksync_era"},
+                   {mojom::kGoerliChainId, "eth_goerli"}});
+
+  return *blockchains;
+}
+
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_constants.cc:131]: If possible please add some link to docs with ankr chain_id strings

File: components/brave_wallet/browser/brave_wallet_service.cc [hunk 8]
```diff
@@ -2192,4 +2192,16 @@ void BraveWalletService::DiscoverEthAllowances(
...
+void BraveWalletService::GetAnkrSupportedChainIds(
+    GetAnkrSupportedChainIdsCallback callback) {
+  const auto& blockchains = GetAnkrBlockchains();
+
+  std::vector<std::string> chain_ids;
+  for (const auto& entry : blockchains) {
+    chain_ids.push_back(entry.first);
+  }
+
+  std::move(callback).Run(chain_ids);
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_service.cc:2204]: std::move(chain_ids)

File: components/brave_wallet/browser/asset_discovery_task.cc [hunk 9]
```diff
@@ -78,16 +79,29 @@ void AssetDiscoveryTask::DiscoverAssets(
                                   ? eth_it->second
                                   : std::vector<std::string>();
 
+  const auto ankr_blockchains = GetAnkrBlockchains();
```
Comment (supermassive) [inline components/brave_wallet/browser/asset_discovery_task.cc:82]: const auto& ankr_blockchains

**Review by josheleonard — APPROVED** [review]
Frontend ++

File: components/brave_wallet/browser/asset_discovery_task.cc [hunk 10]
```diff
@@ -78,16 +79,29 @@ void AssetDiscoveryTask::DiscoverAssets(
...
+
   // Concurrently discover ETH ERC20s on our registry, Solana tokens on our
   // Registry and NFTs on both platforms, then merge the results
   const auto barrier_callback =
       base::BarrierCallback<std::vector<mojom::BlockchainTokenPtr>>(
-          3,
+          4,
           base::BindOnce(&AssetDiscoveryTask::MergeDiscoveredAssets,
                          weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
   // Currently SPL tokens are only discovered on Solana Mainnet.
```
[the remark this answers is not in the document]
Comment (onyb) [reply components/brave_wallet/browser/asset_discovery_task.cc]: I just tested that it works (no errors in console/logs). I still added a check for `ankr_evm_chain_ids` being empty in addition to Ankr feature being enabled.

File: components/brave_wallet/common/brave_wallet.mojom (hunk 4 above)
Comment (onyb) [reply components/brave_wallet/common/brave_wallet.mojom]: https://github.com/brave/brave-core/pull/20342/commits/8113ad47f2594b2a9399cfc3b42d31ea060d8a94

File: browser/brave_wallet/brave_wallet_service_unittest.cc [hunk 11]
```diff
@@ -3172,4 +3172,30 @@ TEST_F(BraveWalletServiceUnitTest, GenerateReceiveAddress_Btc) {
...
+            mojom::kPolygonZKEVMChainId,
+            mojom::kRolluxMainnetChainId,
+            mojom::kSyscoinMainnetChainId,
+            mojom::kZkSyncEraChainId,
+            mojom::kGoerliChainId};
+        EXPECT_EQ(chains.size(), expected_chains.size());
+        for (const auto& chain : chains) {
+          EXPECT_TRUE(base::Contains(expected_chains, chain));
+        }
+      }));
```
Comment (supermassive) [inline browser/brave_wallet/brave_wallet_service_unittest.cc:3198]: nit: `EXPECT_THAT(chains, testing::UnorderedElementsAreArray(expected_chains));`
or `ElementsAreArray` if we need ordering

**Review by supermassive — COMMENTED** [review]
wallet core lgtm

**Review by supermassive — APPROVED** [review]

File: browser/brave_wallet/brave_wallet_service_unittest.cc (hunk 11 above)
Comment (onyb) [reply browser/brave_wallet/brave_wallet_service_unittest.cc]: https://github.com/brave/brave-core/pull/20342/commits/dbb5388879448a30cc2daaadbcc50854bd937e00

**Review by StephenHeaps — APPROVED** [review]

**Pull request merged**